### ISSUE TITLE
SSH key fingerprint differs from the one provided by GPG

### ISSUE THREAD

**Issue by desultory-zz** [rupor-github/win-gpg-agent#34]
I'm pretty confused here, when I run "gpg --export-ssh-key z" it outputs:
`ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIAZHFjxkyRTgqb2tMbiEvEkNcg79knqZJfMBjEBtrMq+ openpgp:0x967DD989`

When I ran ssh-copy-id with this agent running and it installed two keys, but the relevant one was:
`ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAINliIwfA09BAc8G0e4AIYn6cYbjFUimfDuUKPcvgLQLP cardno:xx xxx xxx`

I don't see why it's seemingly generating two different fingerprints but I may be missing something.  I think the other key that it added was my PIV key which is 384 bits, I just deleted this key from the ssh authorized keys, and am reasonably certain this key has to be from my yubikey because the serial number matches and it logs me in, I also have no other keys in my keyring currently. I just have no idea how or why I'm getting different supposed ssh keys.  

I'm not really complaining that much because this works unlike the windows openssh agent, pageant, or anything else I've tried, but I'm kinda concerned that this is deriving different keys, based on the description of this project, that doesn't really seem like something it should even do.

**Comment by desultory-zz** [comment]
I've told gpg to force using my authentication subkey and it still resulted in the Mq+ key being used.  I've verified that the keys in `gpg --card-status` and `gpg --edit-key z` are the same, so I don't know why this discrepancy exists.

**Comment by desultory-zz** [comment]
GPG itself gives me the expected ssh key with `gpg --export-ssh-key {subkey}` or `gpg --export-ssh-key {key}` as the only key I have in my keyring has a single authentication key. My Yubikey has a piv certificate as well as GPG certificates.  Your tool is able to use both for ssh auth, but for some reason, it is deriving a different ssh key from my gpg keys than gpg itself which makes no sense to me.  I don't know how your wrapper derives the ssh public key, but it seems to differ from how GPG does it.

**Comment by rupor-github** [comment]
@desultory-zz Tool does not derive anything - it simply does not know how. All it does it serves as a set of pipes (see picture in readme) connecting different toolsets. It has no idea what gpg does, what is a key etc. It simply controls `gpg-agent` lifetime on Windows side and provides various pipes. If you see 2 different keys - it means that you installation is confused somehow and you are looking in 2 different places (which is very easy - say you could by accident to have gpg-agent inside and outside WSL) of you could have gpg-agent and ssh-agent in multiple places or some other misconfiguration. This is the reason why I am deprecating this - to much unnecessary complexity and to many places to look at.

**Comment by desultory-zz** [comment]
You're deprecating this tool? 

I'm still not sure what is going on and it seems concerning to me.  
Can you explain how it gets the ssh keys for the ssh agent? My yubikey is using gpg and piv certificates.  When I use the gpg tool to get the ssh authentication public key from my smart card, it gives me a different key from what I'm seeing in my SSH agent when your tool is running.  When it is not running, I see no keys associated with any smart card.

The fact that it's showing a different sha265 hash from the one that GPG gives me is very concerning, and I'd again like to ask how you think this is even happening.  I don't know how much more debug info I can provide to clarify this issue but I feel like I'm just repeating myself at this point.  Do you not see the issue with two keys apparently being derived from a smart card?  The serial keys on the outputs match and everything so this really makes no sense to me.

In the past I've use the ssh key derived from GPG just like I described to add my SSH keys to a server and log in, but as you know, this doesn't really work that well with windows 11, so again, I'm wondering why the auth keys I get when running your tool are somehow different from the ones from gpg.   It makes no sense to me but is clearly and repeatably happening.

**Comment by rupor-github** [comment]
If you are really concerned - **why would not you investigate your setup**? I cannot do it for you! 

Project's readme explains what this tool does in details: it controls life-time of "gpg-agent" and extends its piping. Period - it does not use other gpg tools. It knows nothing about how you prepare your keys. In readme there is a diagram of resulting piping - have you looked at it and read it through?

Can you describe or understand you complete setup in detail? When you are saying " I use the gpg tool" - which gpg tool is that? Exact version? Is it Windows one? How was it installed - using winget? chcolatey? scoop? something else? How was it configured and where is configuration file? Can you turn on gpg logging using configuration (I outlined in projects readme how to do that) and look at or share the results? Where gpg-agent you are trying to control and use keeps its sockets - exact paths? 

When you are using "gpg --export-ssh-key z" it does not tell where you "gpg-agent" is. is it running? Is it possible that "ssh-agent" is also running? Can it be that you have one key in "gpg-agent" and another in "ssh-agent"? If "ssh-agent" is in use (as I suspect is the case) - which "ssh-agent"? WIndows one? Cygwin one? 

What do you mean by "as you know that does not really work with Windows 11"? I do not know - everything worked with Windows 11 on my side perfectly for a very long time.

As you can see when I am saying that I do not really understand what is going on from your questions - I really do not understand. So please, start making sense ore let's close this issue as useless.

**Comment by desultory-zz** [comment]
If you wanted to know my setup you could have asked, I'm on Windows 11 and I'm using gpg version 2.3.6 installed from gnupg (not gpg4win).  I'm using the windows ssh client but have also tested with wsl2.

Yes, I have looked at the diagram and made an attempt to understand it, it seems clear which is why I'm confused about this unexpected behavior.  I've been asking for help and trying to describe my setup but you're not asking for any specifics.  I'm trying to provide you with them now since you apparently want them.  I have disabled the windows ssh agent and don't have another one running to my knowledge.  Once could be running in WSL but I'm largely not using that for testing since I'm trying to use my Yubikey here and its easiest to do that through windows directly.

**Comment by desultory-zz** [comment]
I've done my best to investigate my setup, and everything on my end is giving me the same hash but once I start running your tool, I get different hashes.  This is extremely confusing behavior to me

**Comment by rupor-github** [comment]
The burden is on you - not on me. There are too many variables here and I clearly stated that I do not know what to ask for, If "my tool" does not provide what you need - do not use it.  To the best of my knowledge Windows (and OpenSSH tools which come with it) do not have "ssh-copy-id" script. This means that you have installed something else - may be git (which comes with its own Cygwin based copy of SSH) or something else... In cannot possibly ask about something I have no idea about.

**Comment by desultory-zz** [comment]
I'd like to provide you with whatever information you'd like, you haven't really asked for anything specific.  And yes, I did run ssh-copy-id from WSL2, but I'm also seeing these keys when I run `ssh-add -L` from windows or WSL.  From windows, it no longer shows any keys in `ssh-add -L` when I stop running your tool.

**Comment by desultory-zz** [comment]
maybe the burden is on me but it seems to me there is an issue with your software or at least how I'm using it when I get no ssh keys in my agent when it isn't running, but suddenly I get ones that are correlated with my yubikey when I run your program but their hash differs from what gpg itself tells me.  I also have no idea why it's getting my yubikey PIV keys as well, I've never directly used this functionality with GPG and only configured that with yubikey tools but I guess your tool has this functionality. I'm trying to be as helpful as I can here with providing information but you really don't seem to want to help.

**Comment by desultory-zz** [comment]
Ah, forgot this comment `If gpg authors do not want to deal with this - why should I?`

**Comment by rupor-github** [comment]
I am sorry that I cannot help you. I am sorry that you are not getting what you want from "my tool". My tool - `win-gpg-agent` does not have ANY functionality by itself (not even sure what PIV keys are) and I tried to explain it multiple times - all it does is provided by `gpg` itself. `win-gpg-agent` and `gpg` both have debug logging which could be looked at, if necessary, which I also mentioned several times. There are people who are using Yubikey with "my tool" successfully from the beginning - but I am not one of them (never had it). I have no intention of continuing development of this tool precisely because environment it run in is too complicated to be reliable - and I wrote this as first paragraph of my documentation.

**Comment by rupor-github** [comment]
I hope you feel better now, after writing all this! Did you even read what I was answering? My first answer was "I honestly do not understand the issue, sorry." - literally...

**Comment by desultory-zz** [comment]
The first thing I said was "I'm pretty confused here".  I did my best to help you understand the issue but you don't seem to want to try to understand it and would rather tell me to figure it out while not providing me with real advice.  I did read everything you said, and have reread it in the pursuit of trying to resolve this issue.  I've asked similar questions elsewhere and the response was very insightful information: https://superuser.com/questions/1747917/unable-to-use-windows-openssh-agent-with-gpg-agent-and-yubikey/1747919#1747919 

I've tuned the gpg agent debugging to the max and it's not telling me anything useful about the situation, and often just doesn't log anything when I'm trying to test things.  It would be amazing if you, the creator of this "set of pipes" could respond giving me information like in this superuser responder who - seeing I was confused - helped me understand what I was dealing with instead of telling me to google it, as I have, and there is very little well written documentation about the mess that is ssh on windows
